nexus/src/app/instance_network.rs

```diff
@@ -361,29 +368,55 @@ impl super::Nexus {
             })
             .map(|(_, ip)| ip)
         {
-            retry_until_known_result(log, || async {
-                dpd_client
-                    .ensure_nat_entry(
-                        &log,
-                        target_ip.ip,
-                        dpd_client::types::MacAddr {
-                            a: mac_address.into_array(),
-                        },
-                        *target_ip.first_port,
-                        *target_ip.last_port,
-                        vni,
-                        sled_ip_address.ip(),
-                    )
-                    .await
-            })
-            .await
-            .map_err(|e| {
-                Error::internal_error(&format!(
-                    "failed to ensure dpd entry: {e}"
-                ))
-            })?;
+            // For each external ip, add a nat entry to the database
+            self.ensure_nat_entry(
+                target_ip,
+                sled_address,
+                &network_interface,
+                mac_address,
+                opctx,
+            )
+            .await?;
         }
 
+        // Notify dendrite that there are changes for it to reconcile.
+        // In the event of a failure to notify dendrite, we'll log an error
+        // and rely on dendrite's RPW timer to catch it up.
+        if let Err(e) = dpd_client.ipv4_nat_trigger_update().await {
+            error!(self.log, "failed to notify dendrite of nat updates"; "error" => ?e);
+        };
+
+        Ok(())
+    }
+
+    async fn ensure_nat_entry(
+        &self,
+        target_ip: &nexus_db_model::ExternalIp,
+        sled_address: Ipv6Net,
+        network_interface: &sled_agent_client::types::NetworkInterface,
+        mac_address: macaddr::MacAddr6,
+        opctx: &OpContext,
+    ) -> Result<(), Error> {
+        match target_ip.ip {
+            IpNetwork::V4(v4net) => {
+                let nat_entry = Ipv4NatValues {
+                    external_address: Ipv4Net(v4net).into(),
+                    first_port: target_ip.first_port,
+                    last_port: target_ip.last_port,
+                    sled_address: sled_address.into(),
+                    vni: DbVni(network_interface.vni.clone().into()),
+                    mac: nexus_db_model::MacAddr(
+                        omicron_common::api::external::MacAddr(mac_address),
+                    ),
+                };
+                self.db_datastore
+                    .ensure_ipv4_nat_entry(opctx, nat_entry)
+                    .await?;
+            }
+            IpNetwork::V6(_v6net) => {
+                // TODO: implement handling of v6 nat.
```

Comment:
If this is unimplemented, we should throw an error rather than silently doing nothing